### DISCUSSION TITLE
Privacy and oidc

### DISCUSSION THREAD

**Discussion by idc77** [ent/ent#2097] [Q&A]
I have read https://entgo.io/docs/privacy
and I'm like having these wires in my hand but I don't know where to connect them.

Request comes in, has a header "Authorization":"Bearer &lt;token&gt;"
In this token there are claims
```go
// simplified, condensed claims struct
type Claims struct {
	Sub         string `json:"sub,omitempty"`
	RealmAccess struct {
		Roles []string `json:"roles,omitempty"`
	} `json:"realm_access,omitempty"`
	PreferredUsername string `json:"preferred_username,omitempty"`
}
```
oidc client autoconfiguration, signature checking is done by this package
https://github.com/XenitAB/go-oidc-middleware/blob/main/internal/oidc/oidc.go

It has no support for ent because that's a different perspective. Still, parts of it could be ripped out or maybe a PR for ent could be contributed.

So I'm now sitting here trying to make sense of it all.
I have a ClaimsFromRequest() function which is bound to gin.

```go
func ClaimsFromRequest(cx *gin.Context) *Claims {
	claimsValue, found := cx.Get("claims")
	if !found {
		fmt.Println("!found")
		return nil
	}

	claims, ok := claimsValue.(map[string]interface{})
	if !ok {
		fmt.Println("!ok")
		return nil
	}

	c := new(Claims)
	c.Sub, _ = claims["sub"].(string)
	c.PreferredUsername, _ = claims["preferred_username"].(string)
	c.RealmAccess, _ = claims["realm_access"].(struct {
		Roles []string `json:"roles,omitempty"`
	})
	return c
}
```

But ent is http agnostic.
It knows of context.Context

Does not compute, does not compute, does not compute

**Comment by idc77** [comment]
Thinking about this with a fresh brain,
and taking the example into consideration

The example gets a context passed where it creates a Visitor out of it.
The privacy .. uh.. function takes this Visitor to decide various stuff.

So,
```
[oidc] ---pass context---> [package creating visitor struct from context] ---pass visitor--> [rule package doing stuff with visitor struct]
```
I'll get back to this.

**Reply by aniforprez** [reply]
This is pretty much what I have in mind too. I'm only just getting started with auth and such but I think this is the flow that needs to be implemented to pass on the current user to the rest of the application and ent
